delfin/drivers/dell_emc/power_store/rest_handler.py

```diff
@@ -0,0 +1,958 @@
+                'name': appliance.get('name'),
+                'storage_id': storage_id,
+                'native_storage_pool_id': appliance_id,
+                'status': constants.StoragePoolStatus.NORMAL,
+                'storage_type': constants.StorageType.BLOCK,
+                'total_capacity': total_capacity,
+                'used_capacity': used_capacity,
+                'free_capacity': total_capacity - used_capacity
+            }
+            list_pool.append(pool_result)
+        return list_pool
+
+    def get_volumes(self, storage_id):
+        list_volume = []
+        volumes = self.rest_call(self.REST_VOLUME_URL)
+        for volume in volumes:
+            snapshot_type = volume.get('type')
+            if consts.CHARACTER_SNAPSHOT == snapshot_type:
+                continue
+            volume_type = consts.VOLUME_TYPE_MAP.get(
+                snapshot_type, constants.VolumeType.THIN)
+            status = consts.VOLUME_STATUS_MAP.get(
+                volume.get('state'), constants.StorageStatus.UNKNOWN)
+            volume_id = volume.get('id')
+            total_capacity = volume.get('size')
+            used_capacity = self.get_volume_used_capacity(
+                volume_id, volume_type, total_capacity)
+            volume_result = {
+                'name': volume.get('name'),
+                'storage_id': storage_id,
+                'description': volume.get('description'),
+                'status': status,
+                'native_volume_id': volume_id,
+                'native_storage_pool_id': volume.get('appliance_id'),
+                'wwn': volume.get('wwn') if volume.get('wwn') else '',
+                'type': volume_type,
+                'total_capacity': total_capacity,
+                'used_capacity': used_capacity,
+                'free_capacity': total_capacity - used_capacity
+            }
+            list_volume.append(volume_result)
+        return list_volume
+
+    def get_volume_used_capacity(self, volume_id, volume_type, used_capacity):
+        if volume_type == constants.VolumeType.THICK:
+            return used_capacity
+        data = {'entity': consts.SPACE_METRICS_BY_VOLUME,
+                'entity_id': volume_id}
+        volumes_spaces = self.rest_call(self.REST_GENERATE_URL, data, 'POST')
+        if volumes_spaces:
+            volumes_space = \
+                volumes_spaces[consts.DigitalConstant.MINUS_ONE_INT]
+            used_capacity = volumes_space.get('logical_used')
+        return used_capacity
+
+    def get_disks(self, storage_id):
+        disk_list = []
+        hardware_list = self.rest_call(self.REST_HARDWARE_URL)
+        for hardware in hardware_list:
+            lifecycle_state = hardware.get('lifecycle_state')
+            if consts.CHARACTER_DRIVE != hardware.get('type') or \
+                    lifecycle_state == consts.CHARACTER_EMPTY:
+                continue
+            extra_details = hardware.get('extra_details')
+            capacity = ''
```

Comment:
The capacity of disk can't be a string type, if it's None, we should record the log and ignore this disk.

Reply:
It has been modified.

 capacity = None
            firmware = ''
            physical_type = constants.DiskPhysicalType.UNKNOWN
            if extra_details:
                firmware = extra_details.get('firmware_version')
                drive_type = extra_details.get('drive_type')
                if drive_type in consts.DiskType.ALL:
                    continue
                physical_type = consts.DISK_PHYSICAL_TYPE.get(
                    drive_type, constants.DiskPhysicalType.UNKNOWN)
                capacity = extra_details.get('size')
            hardware_name = hardware.get('name')
            if not capacity:
                LOG.error("disk capacity is null: %s", hardware_name)
                continue